W-12043308-Add-expand-feature-CP (#103)
mulesoft/docs-apikit: modules/ROOT/pages/apikit-odatav4-expand-feature.adoc

* Use `GET https://example.com/api/Managers?$filter=ManagerID in (id1, id2, ..., idN)` to retrieve a list of related managers.
* Write a small program to merge the lists based on join criteria.

If using the `$expand` feature, requests are `GET https://example.com/api/Employees?$expand=Manager` to retrieve a list of employees with their `Manager` navigation property expanded.

== The `generic expand` Feature

The `generic expand` feature enables you to reuse existing flow implementations as a starting point for expanding requests. Using this feature, you can focus on the relevant data retrieval operations while the module performs the necessary steps to ensure that the `$expand` fields complete.

An application with the `generic expand` feature enabled works as any other OData Mule application. However, during the flow execution, once the module retrieves information for the main entity set, it performs additional internal requests gathering the information for any `expand` field. If there are errors during the expansion operation, the consumer can configure the final behavior, or the consumer can ignore it with a `null` value.

== The `generic expand` Operation

> The `generic expand` feature dispatches additional queries to the application when the original flow completes. Because this is done based on the application `EDM`, extra annotations may be needed in order to add enough context to the module:

Author: ...you need to make extra annotations to add enough context to the module:

> . The `EntitySet` where the expansion takes place should have a `NavitationPropertyBinding` for every expandable `NavigationProperty` of its `Entity`. +

Author: @iglosiggio, please, can you confirm the correct name?:
`NavitationPropertyBinding` (with a "t")
`NavigationPropertyBinding` (with a "g")
Thanks!

> = Implement $expand Feature

Reviewer: ```suggestion
= Implement the $expand Feature
```

> = Implement the $expand Feature

Reviewer 2: This title sound a bit weird. What about "Implement $expand using _generic expand_"?

Author: The title "Implement $expand using generic expand" assumes that you know the meaning of "$expand" and the meaning of "generic expand". In a perfect world, we would have:

- APIkit
  - APIkit for OData
    - Create an OData v4 API with APIkit
      - The $expand feature
      - The general expand feature
      - Implement $expand using generic expand

Actually, the editor suggested that we should split this article into two or three when we rearrange the APIkit nav tree. So this may be a great idea for the near future. 
What do you think, @iglosiggio? Thanks!

Reviewer 2: All in for splitting the article at some future point 🚀

Author: Absolutely!
I'll be contacting you so you can help me with such tasks! hahaha
Thanks!

> The `$expand` system query option reduces the amount of incoming network traffic and makes APIs easier to consume. Because consumers request only inline representations of the related entities of query results, the returned information avoids multiple network roundtrips. This benefits metered and latency-sensitive consumers.

Author: @iglosiggio, please, can you clarify if we are talking about the APIkit consumer (AKA, Mule Dev) or the API consumer (who consumes the API developed by Mule Dev)? Thanks!

Reviewer 2: The person and/or software that consumes the API created by the Mule Developer.

Author: OK! Great! Thanks!

> `$expand` is a system query option that reduces the amount of incoming network traffic and makes APIs easier to be consumed. Consumers only request inline representations for the related entities of query results so only relevant information is provided avoiding multiple network roundtrips, which is useful for metered and latency-sensitive consumers.

Reviewer: ```suggestion
The `$expand` system query option reduces the amount of incoming network traffic and makes APIs easier to consume. Consumers only request inline representations for the related entities of query results so only relevant information is provided avoiding multiple network roundtrips, which is useful for metered and latency-sensitive consumers.
```

> `$expand` is a system query option that reduces the amount of incoming network traffic and makes APIs easier to be consumed. Consumers only request inline representations for the related entities of query results so only relevant information is provided avoiding multiple network roundtrips, which is useful for metered and latency-sensitive consumers.

Reviewer: I tried to tighten up the wording and make shorter sentences:

The `$expand` system query option reduces the amount of incoming network traffic and makes APIs easier to consume.
Because consumers request only inline representations of the related entities of query results, the returned information avoids multiple network roundtrips. This benefits metered and latency-sensitive consumers.

Author: Thanks for your great collaboration!

> From APIkit for OData 1.3.0 onwards, this module supports automatic routing for ease of implementing the `$expand` system query option in the most common cases. Although this feature requires more preparation and configuration, it enables the reuse of already implemented flows when processing the `$expand` requests.

Reviewer 2: "when processing the $expand requests" sounds weird to me (I'm not an english native so idk)

$expand requests are not inherently different for other requests. `$expand` is an option that can be added to any given request to ask for special processing of the indicated relationships.

$expand can be used in conjunction with almost any other system query option ($filter, $skip, $top, etc)

Author: It does not seem weird to me or to the editor, @iglosiggio.
Would it be more accurate to say "Although this feature requires more preparation and configuration, it enables the reuse of already implemented flows when including the `$expand` requests."?

Reviewer 2: Lets keep it as-is. :)

> From APIkit for OData 1.3.0 onwards, this module supports automatic routing for ease of implementing the `$expand` system query option in the most common cases. Although this feature requires more preparation and configuration, it enables the reuse of already implemented flows when processing the `$expand` requests.

Reviewer: Optional: 
already implemented flows> implemented flows
or already implemented  flows> existing flows

Author: Great catch!
Thanks!

> 

Reviewer: on > about (this was a Kim directive)

> From APIkit for OData 1.3.0, this module supports automatic routing for ease of implementing the `$expand` system query option in most common cases. Though this feature requires more preparation and configuration, it enables the reuse of already implemented flows when processing the `$expand` requests.

Reviewer: I am not sure what you mean by "From APIkit for OData 1.3.0. Do you mean "For APIkit..."?

in most common cases > in the most common cases

Though > Although

Reviewer: Maybe we should discuss the wording for this. I can't really tell what you are trying to say here.

Author: APIkit for OData 1.3.0. is the first version that supports this feature. It is not supported by previous APIkit versions.
Let's talk about this paragraph in our meeting.

Author: From APIkit OData 1.3.0 onwards

> For more information on the system query option, see http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_SystemQueryOptionexpand[the OASIS specification^].

Reviewer: on > about (this was a Kim directive)

Author: Kim's word! 
Thanks!

== Requirements

> 

Reviewer: Remove periods from the following list.

Author: OK! Thanks!

- APIkit for OData Module (version 1.3.0).

> - An `<request-entity-collection-listener/>` implementation for each expandable entity.

Reviewer: An > A because <request-entity-collection-listener/> is before "implementation."

Author: OK! Thanks!

- Support for the `in` filter on the `<request-entity-collection-listener/>` flows.

> == `$expand` Feature Example 

Reviewer: ```suggestion
== `$expand` Feature Example 

In this example:
```

Author: Thanks!

* The `Employee` entity has a related `Manager` entity.
* The `Employees` entity set is a collection of `Employee` entities.
* The `Managers` entity set is a collection of `Manager` entities.

> * API location is `https://example.com/api/`.

Reviewer: Add "The," since you have it for the other bullet entries.

Author: Right you are! Thanks!

> Given this setup, requests to retrieve employees with their managers are:

Reviewer: To avoid the passive:

Given this setup, the following requests retrieve employees and their managers:

Author: OK! Thanks!

> * `GET https://example.com/api/Employees` to retrieve a list of employees.

Reviewer: This needs to start with a verb to be consistent with the next entry. The `GET` in `GET https://example.com/api/Employees` is part of the command, so it is not parallel. I would say:

 "Use `GET https://example.com/api/Employees` to...

Global for similar wording situtations

Author: Sure! Thanks!

* Gather the `Manager` IDs through the `ReferentialConstraint` of their relationship.
* `GET https://example.com/api/Managers?$filter=ManagerID in (id1, id2, ..., idN)` to retrieve a list of related managers.
* Write a small program to merge the lists based on join criteria.

If using the `$expand` feature, requests are:

> * `GET https://example.com/api/Employees?$expand=Manager` to retrieve a list of employees with their `Manager` navigation property expanded.

Reviewer: This should not be a bullet since it is only one item.

Author: OK! I'll change it. Thanks!

> If using the `$expand` feature, requests are `GET https://example.com/api/Employees?$expand=Manager` to retrieve a list of employees with their `Manager` navigation property expanded.

Reviewer: "requests are" implies multiple requests, but you are showing only one

Author: Right you are!
This only request replaces many other requests, but it is still only one!
Thanks!

== The `generic expand` Feature

> The `generic expand` feature offers the possibility to reuse existent flow implementations as a starting point for expanding requests. This feature enables a Mule developer to focus on the relevant data retrieval operations while the module performs the necessary steps to ensure the `$expand` fields get completed.

Reviewer: ```suggestion
The `generic expand` feature enables you to reuse existing flow implementations as a starting point for expanding requests. Using this feature, you can focus on the relevant data retrieval operations while the module performs the necessary steps to ensure that the `$expand` fields complete.
```

Reviewer: For the text above, I changed the second sentence to avoid using "enables" twice in a row.

Author: Thanks!

> An application with the `generic expand` feature enabled works as any other OData Mule application. However, during the flow execution, once the module retrieves information for the main entity set, it performs additional internal requests gathering the information for any `expand` field. If there are errors during the expansion operation, the consumer can configure the final behavior, or the consumer can ignore it with a `null` value.

Author: @iglosiggio, please, can you clarify if we are talking about the APIkit consumer (AKA, Mule Dev) or the API consumer (who consumes the API developed by Mule Dev)? Thanks!

Reviewer 2: The mule developer is the person responsible for configuring error handling behaviour.

Author: OK! Thanks for the clarification!

> An application with the `generic expand` feature enabled works as any other OData Mule application. However, during the flow execution, once the module retrieves information for the main entity set, it performs additional internal requests gathering the information for any `expand` field. If there are errors during the expansion operation, the consumer can configure the final behavior, or the consumer can ignore it with a `null` value.

Reviewer 2: "The developer can configure the final behaviour: either to fail the request with an error or to replace the failed expansion with a `null` value."

Those are the two options: Big explosive failure or stealthy null value.

Author: hahahaha!
Thanks for the clarification!

> An application with the `generic expand` feature enabled works as any other OData Mule application. However, during the flow execution, once the module retrieves information for the main entity set, it performs additional internal requests gathering the information for any `expand` field. If there are errors during the expansion operation, you can configure the final behavior. You can fail the request with an error, or you can replace the failed expansion with a `null` value.

Reviewer: ```suggestion
An application with the `generic expand` feature enabled works like any other OData Mule application. However, during the flow execution, once the module retrieves information for the main entity set, it performs additional internal requests to gather the information for any `expand` field. If there are errors during the expansion operation, you can configure the final behavior. You can fail the request with an error, or you can replace the failed expansion with a `null` value.
```

Reviewer: Suggestion: 

Change "If there are errors during the expansion operation, you can configure the final behavior." to the following text. The current text is awkward.

You can configure how the software handles errors during the expansion operation: 
* Fails the request with an error.
* Replaces the failed expansion with a `null` value.

Author: I like your suggestion! 
Thanks!

> From the point of view of a Mule developer, an application with the `generic expand` feature enabled works as any other OData Mule application. The difference in the flow execution is that once the information for the main entity set is retrieved, the module will perform additional internal requests gathering the information for any `expand` field. If errors occur during expansion operation, the final behavior is configurable via the error to the consumer, or it can be ignored, resolving it with a `null` value.

Reviewer: An application with the `generic expand` feature enabled works like any other OData Mule application. However, once the module retrieves information for the main entity set, it performs additional internal requests gathering the information for any `expand` field.

You can write the last sentence more actively. I was not able to do so because I am not sure what you mean by "error to the consumer."

Author: I'll take your suggestion and make some changes since I feel some info would be lost otherwise.
And I'll rewrite the "error" part because it is not clear with this wording.
Thanks!

== `generic expand` Operation

> `generic expand` dispatches additional queries to the application when the original flow is completed. Because this is done based on the application `EDM`, extra annotations may be needed in order to add enough context to the module:

Reviewer: ```suggestion
`generic expand` dispatches additional queries to the application when the original flow completes. Because this is done based on the application `EDM`, extra annotations may be needed in order to add enough context to the module:
```

> `generic expand` dispatches additional queries to the application when the original flow is completed. Because this is done based on the application `EDM`, extra annotations may be needed in order to add enough context to the module:

Reviewer: I would say "The `generic expand` operation since there is also a `generic expand` feature.

Author: OK! Thanks!

The `EntitySet` where the expansion takes place should have a `NavitationPropertyBinding` for every expandable `NavigationProperty` of its `Entity`. +

> This is used to select which `<entity-collection-source/>` will be executed when expanding that property.

Reviewer: ```suggestion
This is used to select which `<entity-collection-source/>` is executed when expanding that property.
```

. Each expandable `NavigationProperty` has a `ReferentialConstraint` attached. +

> This creates the inner query, where the value of `Property` is matched against the value of `ReferencedProperty`.

Reviewer 2: **Creates** sounds weird. The `EDM` doesn't really create anything.

Maybe "This is used by the internal requests, where ..."

Author: Great! Thanks for the clarification!

> Internal requests use this, where the value of `Property` is matched against the value of `ReferencedProperty`.

Reviewer: Not sure what "where" means here. Does it mean "when" or is the following wording correct?

Internal requests use this to match the value of `Property` to the value of `ReferencedProperty`.

Author: "When" is the accurate one!
Thanks!

> This is used to create the inner query where the value from the `Property` will be matched against the value of the `ReferencedProperty`.

Reviewer: ```suggestion
This creates the inner query, where the value of `Property` is matched against the value of `ReferencedProperty`.
```

=== Flow Execution

> . The main flow is executed and completed resulting in a collection of `entities`.

Reviewer: ```suggestion
. The main flow executes and completes, resulting in a collection of `entities`.
```

Author: Thanks!

. The OData module gathers all the properties that need expansion.
. For each `NavigationProperty`:
.. Get the `source property` from its `ReferentialConstraint`.
.. Get the `referenced property` from its `ReferentialConstraint`.
.. Get the `destination entity set` from its `NavigationPropertyBinding`.

> .. For each of the `entities` get the resolved `source property`. Put all these values on a `source property list`.

Reviewer: ```suggestion
.. For each of the `entities` get the resolved `source property`. Add all of these values to a `source property list`.
```

Author: Why emphasize "all of these" instead of just "these"?

Author: Add these values... is OK

> .. Execute the `internal` request by invoking the `<entity-collection-source/>` for `destination entity set` with `$filter=referenced property` in `(...source property list)`.

Reviewer 2: I don't think internal has to be highlighted here. It's not a specification thing nor a name for something that will be referenced later.

Author: OK! Thanks for the clarification!

=== Error Handling and Error Propagation

> Error handling for OData v4 applications is local to each flow. Whether you use `generic expansion` or not. Error handling code for a given `EntitySet` will belong to the flow implementations for that entity set, both for normal requests and for the internal requests observed during expansion processing.

Reviewer: ```suggestion
Error handling for OData v4 applications is local to each flow, whether you use `generic expansion` or not. Error handling code for a given `EntitySet` belongs to the flow implementations for that entity set, both for normal requests and for the internal requests observed during expansion processing.
```

> Handling errors during the expansion process is not supported, because the expansion process runs once the execution of the flow is finished. There are two options available for handling errors:

Reviewer: ```suggestion
Handling errors during the expansion process is not supported, because the expansion process runs after the execution of the flow finishes. There are two options available for handling errors:
```

> - *Ignore errors during expansion:* Errors during generic expansion will be ignored. +

Reviewer: ```suggestion
- *Ignore errors during expansion:* Errors during generic expansion are ignored. +
```

> By default, the expansion of collections and single objects are resolved to empty lists and `null` values respectively.

Reviewer: ```suggestion
By default, the expansion of collections and single objects are resolved to empty lists and `null` values, respectively.
```

Author: Thanks!

> Expansion of collections and single objects will be resolved to empty lists and `null` values respectively (default behaviour).

Reviewer: ```suggestion
By default, the expansion of collections and single objects are resolved to empty lists and `null` values respectively).
```

Author: OK! Thanks!

- *Propagate errors during expansion:* Errors during generic expansion will arise. +

> When an error reaches the OData v4 router, the original request will be answered with an error response.

Reviewer: ```suggestion
When an error reaches the OData v4 router, the original request is answered with an error response.
```

Author: Thanks!

You can configure error handling *per-source* on the *Capabilities* tab.

=== Limitations

* Only the first `ReferentialConstraint` is considered. +
Navigation properties that use compound keys are not supported.
* Expansion is done `field-by-field` instead of `EntitySet-by-EntitySet`. +

> This can cause more internal requests than necessary.

Reviewer: ```suggestion
This can create more internal requests than necessary.
```

Author: Thanks!

> This may cause more internal requests than necessary.

Reviewer: ```suggestion
This can cause more internal requests than necessary.
```

Author: OK! Thanks!

* The expanded entities always have an `<entity-collection-source/>` implementation. +
Expansion using `<entity-source/>` implementations are not executed due to performance concerns.

== Implementation Guidelines

> To implement the `generic expansion` feature, you must annotate each source with the types that it can expand. Enabling `$expand` without restrictions enables API clients to write arbitrarily complex queries, which can create performance issues.

Reviewer: Can we remove "arbitrarily"?

Author: Yes, we can. 
"Complex" provides the meaning as well.
Thanks!

> To implement the `generic expansion` feature, each source has to be annotated with the relevant types that might be expanded from it. Note that enabling `$expand` in a completely unrestricted way will enable API clients to write arbitrarily complex queries, which can create performance issues.

Reviewer: ```suggestion
To implement the `generic expansion` feature, you must annotate each source must with the types that it can expand. Enabling `$expand` without restrictions enables API clients to write arbitrarily complex queries, which can create performance issues.
```

Author: OK! Thanks!

> If you implement `$expand` manually for some relationships, leave such relationships out of the list on the *Expandable Navigation Properties* option. Manual implementations inside the flow will be needed for the navigation properties that don't use the `generic expansion` feature.

Reviewer: ```suggestion
If you implement `$expand` manually for some relationships, leave these relationships out of the list on the *Expandable Navigation Properties* option. Manual implementations inside the flow are needed for the navigation properties that don't use the `generic expansion` feature.
```

Author: OK! Thanks!

=== Implementation Checklist

. Check that your CSDL file and Mule application meet the expected requirements:

> ** Every `NavigationProperty` should have one `ReferentialConstraint`.

Reviewer: You can make this less wordy by changing "should have" to "have."

Author: Right! Thanks!

** Every `EntitySet` should have one `NavigationPropertyBinding` for each `NavigationProperty` of its entity type.
** Every `<entity-collection-source/>` should support the `in` operator for the `$filter` system query option.

> . For each source where you want to enable `generic expand` support:

Reviewer: ```suggestion
. For each source for which you want to enable `generic expand` support:
```

Author: OK! Thanks!

.. Open the *Capabilities* tab

> .. Edit inline *Expandable navigation properties*

Reviewer: ```suggestion
.. Edit inline the *Expandable navigation properties* field.
```

Author: OK! Thanks!

> .. Add the name of each navigation property you want to be expandable using `generic expand`.

Reviewer: ```suggestion
.. Add the name of each navigation property that you want to be expandable by using `generic expand`.
```

Author: Thanks!

.. Depending on your use case, check *Ignore errors on expand* to avoid errors during the generic expansion and to generate errors on the main flow.

> . Verify that your existent flows support the `$expand` feature.

Reviewer: ```suggestion
. Verify that your existing flows support the `$expand` feature.
```

Reviewer: Please add periods to the bullets above since they are full sentences.

Author: Sure! Thanks!
